querybook/webapp/components/EnvironmentsRouter/EnvironmentsRouter.tsx

```diff
@@ -19,199 +12,140 @@ import {
-        if (userEnvironmentNames.has(name)) {
-            for (const environment of environments) {
-                if (
-                    currentEnvironment !== environment &&
-                    name === environment.name
-                ) {
-                    this.loadEnvironment(name);
-                    break;
+const EnvironmentsRouter: React.FC = () => {
+    const [environmentsLoaded, setEnvironmentLoaded] = useState(false);
+
+    const dispatch: Dispatch = useDispatch();
+
+    const environments = useSelector(environmentsSelector);
+    const userEnvironmentNames = useSelector(userEnvironmentNamesSelector);
+    const currentEnvironment = useSelector(currentEnvironmentSelector);
+    const isAdmin = useSelector(
+        (state: IStoreState) => state.user.myUserInfo.isAdmin
+    );
+    const defaultEnvironmentId = useSelector(
+        (state: IStoreState) =>
+            state.user.computedSettings['default_environment']
+    );
+
+    useEffect(() => {
+        dispatch(EnvironmentActions.fetchEnvironments()).finally(() => {
+            setEnvironmentLoaded(true);
+        });
+        dispatch(fetchQueryTranspilers());
+        dispatch(fetchExporters());
+        dispatch(fetchNotifiers());
+    }, [dispatch]);
+
+    const selectEnvironment = useCallback(
+        (name: string) => {
+            if (userEnvironmentNames.has(name)) {
+                for (const environment of environments) {
+                    if (
+                        currentEnvironment !== environment &&
+                        name === environment.name
+                    ) {
+                        dispatch(EnvironmentActions.setEnvironment(name));
+                        break;
+                    }
                 }
             }
-        }
-    }
+        },
+        [dispatch, environments, currentEnvironment, userEnvironmentNames]
+    );
 
-    public componentDidMount() {
-        this.loadEnvironments();
-        this.loadInitialLoadItems();
+    if (!environmentsLoaded) {
+        return <Loading fullHeight />;
     }
 
-    public render() {
-        const { environmentsLoaded } = this.state;
-
-        if (!environmentsLoaded) {
-            return <Loading fullHeight />;
-        }
-
-        const {
-            environments,
-            currentEnvironment,
-            userEnvironmentNames,
-            defaultEnvironmentId,
-            isAdmin,
-        } = this.props;
-
-        let redirectDOM = null;
-        if (userEnvironmentNames.size > 0) {
-            let defaultEnvironment = null;
-            if (defaultEnvironmentId != null) {
-                defaultEnvironment = environments.find(
-                    (env) =>
-                        env.id === Number(defaultEnvironmentId) &&
-                        userEnvironmentNames.has(env.name)
-                );
-            }
-            if (!defaultEnvironment) {
-                defaultEnvironment = environments.find((env) =>
+    let redirectDOM = null;
```

Comment:
have a separate function to `getRedirectDOM()`?

Reply:
sounds good!